man/man8/dbmedit.8
.TH DBMEDIT 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/dbmedit.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: dbmedit.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
dbmedit \- edit the pp database file
.SH SYNOPSIS
.in +.5i
.ti -.5i
dbmedit \%[-v] \%[-d\ database] \%[cmd ...]
.SH DESCRIPTION
.I Dbmedit
is a program to manually edit the \fIdbm\fR(3) database used by PP.
It is intended to be used by the system maintainer for quick
and simple changes to the database.
It can also be used by careful setuid programs to make controlled
changes on behalf of users.
For example, a \fIforwardmail\fR command might use
dbmedit to change a user's entry in the
dbm database after changing the mail forwarding alias file.
.PP
The \-v option may be used to get a verbose description of the
program's activities.
.PP
The \-d option may be used to specify an alternate database.  The
default is given in the PP tailor file.
.PP
If no arguments are given to \fIdbmedit\fR then the program
goes into an interactive mode, and prompts the user for each
command. Otherwise the arguments are taken as one command.
.PP
Commands in \fIdbmedit\fR refer to keys, tables, and values.
Tables are hashed into the database using \fIdbmbuild\fR(8).
The keys appear on the LHS of the tables and the values on the RHS.
In general, only the first occurence of a value for a given
key/table pair is significant.  
For example, the table entries:
.nf
.ta \w'table1:\0\0'u
table1:
	key1: val1      
	key2: val2      
table2:
	key1: val3      
	key1: val4      
.fi
get hashed into the following database entries:
.nf
key1  table1  val1
key1  table2  val3
key1  table2  val4
key2  table1  val2
.fi                        
.PP
The command lines in interactive mode are parsed using the standard
PP string-to-argument routines so the same quoting
and escape conventions are used.
For example, if you want double-quotes or spaces
in the value, they must be
escaped with a backslash or the string must be quoted (for spaces).
.PP
The commands are:
.RS
.TP 4
.I print key [table]
Print the value of the key/table pair.
If the table is omitted, then print the value of any
table entry with this key.
.TP 4
.I add key table value
Add a key/table entry with the given value.
In verbose mode, a warning message is printed if the given
key/table pair already has a value in the database.  
.TP 4
.I delete key [table [value]]
Delete the values for the specified key.  If a table is specified,
delete only the values for the specified key/table pair.  If a value
is also specified, delete only entries for the pair with that value.
It is an error to try to delete something which does not appear in
the database as specified.
.TP 4
.I change key table [oldvalue] newvalue
Change the value of the specified key/table pair to
.I newvalue.
If 
.I oldvalue
is specified, change the entry matching that value.  Otherwise,
change the value of the first occurence or add a new key/table pair
if none already exists.
.TP 4
.I help
Give a brief summary of the commands
.TP 4
.I quit
Exit the program.
.RE
.PP
All commands may be shortened to their first character only.  
If the wrong number of arguments is given to a command, a
``Usage:'' message will be printed.  
All changes are made in real time.
This program may be used while PP processes are running.
There is no temporary copy of the database while editing.
.SH FILES
.nf
.ta \w'ppdbm.dir\0\0'u
ppdbm.dir	The PP database
ppdbm.pag	The PP database
.fi
.SH "SEE ALSO"
dbm(3x), dbmbuild(8), dbmunbuild(8)
.br
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Tools/dbmedit/dbmedit.c
/* dbmedit: edit a dbm file */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Tools/dbmedit/RCS/dbmedit.c,v 6.0 1991/12/18 20:29:41 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Tools/dbmedit/RCS/dbmedit.c,v 6.0 1991/12/18 20:29:41 jpo Rel $
 *
 * $Log: dbmedit.c,v $
 * Revision 6.0  1991/12/18  20:29:41  jpo
 * Release 6.0
 *
 */



#include "head.h"
#include "util.h" 
#include "dbase.h"
#include "sys.file.h"
#include <ctype.h>

extern char *ppdbm;
extern char *dupfpath();
extern char *multcat();

#ifdef GDBM
GDBM_FILE	thedb;
#define store(x,y)	gdbm_store (thedb, x, y, GDBM_REPLACE)
#define fetch(x)	gdbm_fetch (thedb, x)
#else
#ifndef GCC_DBM_OK
#if sparc && defined(__GNUC__)	/* work around bug in gcc 1.37 sparc version */
 #error GCC and dbm do not get along on a sparc - compile with cc.
#endif
#endif
#endif

# ifdef NDBM
# define        fetch(x)        dbm_fetch(thedb, (x))
# define        store(x, y)     dbm_store(thedb, (x), (y), DBM_REPLACE)
# define        delete(x)       dbm_delete(thedb, (x))
DBM     *thedb;
# endif


datum   dcons();

char    ibuf[LINESIZE];
#define MARGS   100
char    *cargv[MARGS];                /* max number of args on a line */
int     cargc;

int     help(), add(), del(), change(), print(), quit();

struct  cmds    {
	char    *cm_name;
	int     (*cm_func)();
	char    *cm_help;
} cmds[] = {
	"help", help,   "Print out help text",
	"add",  add,    "Add an entry",
	"delete", del,  "Delete an entry",
	"change", change, "Change an entry",
	"print", print, "Print an entry",
	"quit", quit,   "Quit the program",
	"h",    help,   0,              /* single character aliases */
	"a",    add,    0,
	"d",    del,    0,
	"c",    change, 0,
	"p",    print,  0,
	"q",    quit,   0,
	"?",    help,   0,
}, *curcmd;

int     verbose;

/*
 * dbm expanded structure
 */

#define MDB     50

int     ndbents;

struct  db      {
	char    *d_table;
	char    *d_value;
} dbs[MDB], *lastdb;

main(argc, argv)
int argc;
char **argv;
{
	int opt;
	extern int optind;
	extern char *optarg;

	sys_init(argv[0]);

	while ((opt = getopt (argc, argv, "vd:")) != EOF) {
		switch (opt) {
		    case 'v':
			verbose++;
			break;
		    case 'd':
			ppdbm = optarg;
			break;
		    default:
			fprintf(stderr, "dbmedit: unknown argument '%c'\n",
				opt);
			exit(NOTOK);
		}
	}
	argv += optind;
	argc -= optind;
	if(!isstr(ppdbm)) {
		fprintf(stderr, "dbmedit: cannot find database path\n");
		exit(NOTOK);
	}
	dbfinit(ppdbm);

	/* if got command line arguments process these instead */
	if(argc > 0){
		avfix(argc, argv);
		if(findcmd() < 0)
			exit(1);
		exit((*curcmd->cm_func)());
	}

	for(;;){
		printf("dbmedit> ");
		(void) fflush(stdout);
		if (gets(ibuf) == NULL)
			exit(0);
		if (!isstr(ibuf))
			continue;
		if ((cargc = sstr2arg(ibuf, MARGS, cargv, " \t")) == 0)
			continue;
		if(findcmd() < 0)
			continue;
		(*curcmd->cm_func)();
	}
}

avfix(ac, av)
int ac;
register char   **av;
{
	register char   **ap;

	for(ap = cargv; ac ; ac--, cargc++)
		*ap++ = *av++;
}

findcmd()
{
	register struct cmds *cp;

	for(cp = cmds ; cp < &cmds[sizeof(cmds)/sizeof(cmds[0])] ; cp++)
		if(lexequ(cp->cm_name, cargv[0]) == 0){
			curcmd = cp;
			return(0);
		}
	fprintf(stderr, "dbmedit: unknown command '%s'\n", cargv[0]);
	return(-1);
}

help()
{
	register struct cmds *cp;

	for(cp = cmds ; cp < &cmds[sizeof(cmds)/sizeof(cmds[0])] ; cp++)
		if(cp->cm_help != 0)
			printf("%s\t\t%s\n", cp->cm_name, cp->cm_help);
}

datum
dcons(value)
char    *value;
{
	datum d;

	d.dptr = value;
	d.dsize = strlen(value)+1;
	return(d);
}

datum
dconskey(value)
char    *value;
{
	char	*p;

	for (p = value; *p; p++)
		if (isupper(*p))
			*p = tolower(*p);
	return(dcons(value));
}

dssplit(str)
register char   *str;
{
	register struct db      *dp;
	static  char    ti[512];

	ndbents = 0;
	str = strcpy(ti, str);
	for(dp = dbs ; dp < &dbs[sizeof(dbs)/sizeof(dbs[0])]; dp++){
		if(*str == 0)
			break;
		dp->d_table = str;
		while(*str && *str != ' ')
			str++;
		*str++ = 0;
		dp->d_value = str;
		while(*str && *str != FS)
			str++;
		ndbents++;
		if(*str == FS)
			*str++ = 0;
	}
	lastdb = dbs + ndbents;
}

add()
{
	datum   d;
	register struct db      *dp;

	if(cargc != 4){
		fprintf(stderr, "Usage: add key table value\n");
		return(9);
	}
	d = fetch(dconskey(cargv[1]));
	if(d.dptr == 0){
		if (verbose)
			printf("Creating new key \"%s\" value \"%s %s\"\n",
						cargv[1], cargv[2], cargv[3]);
		dp = lastdb = dbs;
		goto newalias;
	}
	dssplit(d.dptr);
	for(dp = dbs ; dp < lastdb ; dp++){
		if(lexequ(dp->d_table, cargv[2]) == 0){
			fprintf(stderr,
			  "key \"%s\" already has a value in table \"%s\"\n",
							cargv[1], cargv[2]);
			return(99);
		}
	}
newalias:;
	lastdb++;
	dp->d_table = cargv[2];
	dp->d_value = cargv[3];
	return(dscons());
}

del()
{
	int     i;
	datum   d;
	register struct db      *dp;
	int     changed = 0;

	if (cargc <= 2) {
		fprintf(stderr, "Usage: delete key table|* [table ...]\n");
		return(9);
	}

	d = fetch(dconskey(cargv[1]));
	if (d.dptr == 0) {
		fprintf(stderr, "key \"%s\" not found in database\n", cargv[1]);
		return(99);
	}
	if (strcmp(cargv[2], "*") == 0) {        /* all entries */
		if (verbose)
			printf("Deleteing key \"%s\"\n", cargv[1]);
		goto delall;
	}
	dssplit(d.dptr);
	for(i = 2 ; i < cargc ; i++) {
		for(dp = dbs ; dp < lastdb ; dp++) {
			if(dp->d_table && lexequ(dp->d_table, cargv[i]) == 0) {
				if (verbose)
					printf("deleting \"%s\" from table \"%s\"\n",
							cargv[1], cargv[i]);
				changed++;
				dp->d_table = 0;
				break;
			}
		}
		if(dp >= lastdb){
			fprintf(stderr,
				"\"%s\" does not have a value in table \"%s\"\n",
							cargv[1], cargv[i]);
			continue;
		}
	}
	if(!changed)
		return(99);
	if(lastdb - changed <= dbs) {
		if (verbose)
			printf("All values deleted for \"%s\" - deleting key\n",
								cargv[1]);
delall:;
		if(delete(dcons(cargv[1])) < 0)
			fprintf(stderr, "Delete failed\n");
		return(0);
	}
	return(dscons());
}

change()
{
	datum   d;
	register struct db      *dp;

	if(cargc < 3 || cargc > 4) {
		fprintf(stderr, "Usage: change key table [value]\n");
		return(9);
	}

	d = fetch(dconskey(cargv[1]));
	if(d.dptr == 0) {
		if (verbose)
			fprintf(stderr, "key \"%s\" not found in database\n",
				cargv[1]);
		if (cargc == 4)
			return(add());
		return(0);
	}
	dssplit(d.dptr);
	for(dp = dbs ; dp < lastdb ; dp++)
		if(lexequ(dp->d_table, cargv[2]) == 0){
			if (cargc == 3)
				return(del());
			if (verbose)
				printf("changing \"%s\" to \"%s\"\n",
						dp->d_value, cargv[3]);
			dp->d_value = cargv[3];
			break;
		}

	if(dp >= lastdb){
		if (cargc == 4)
			return(add());
		return(0);
	}
	return(dscons());
}

print()
{
	datum   d;
	register struct db      *dp;

	if(cargc < 2){
		fprintf(stderr, "Usage: print key [table]\n");
		return(9);
	}

	d = fetch(dconskey(cargv[1]));
	if(d.dptr == 0){
		fprintf(stderr, "key \"%s\" not found in database\n", cargv[1]);
		return(99);
	}
	dssplit(d.dptr);
	for(dp = dbs ; dp < lastdb ; dp++)
		if (cargc == 2 || lexequ(dp->d_table, cargv[2]) == 0 ||
		    lexequ("*", cargv[2]) == 0)
			printf("key \"%s\": table \"%s\": value \"%s\"\n",
				cargv[1], dp->d_table, dp->d_value);
	return(0);
}

dscons()
{
	char    tbuf[ENTRYSIZE];
	datum   d;
	register struct db      *dp;
	register char   *p, *q;
	int     changed = 0;

	for(p = tbuf, dp = dbs ; dp <lastdb ; dp++){
		if((q = dp->d_table) == 0)
			continue;
		if(p != tbuf)
			*(p-1) = FS;
		changed++;
		while(*p++ = *q++);
		*(p-1) = ' ';
		for(q = dp->d_value ; *p++ = *q++;);
	}
	if (!changed) {
		fprintf(stderr, "dbmedit: internal error 'no change'\n");
		return(99);
	}
	d = dcons (tbuf);

	if (store(dconskey(cargv[1]), d) < 0) {
		fprintf(stderr, "dbmedit: cannot store new value for '%s'\n", cargv[1]);
	}
	return(0);
}

/*
 * Initialize the dbm file.
 */

dbfinit(filename)
char *filename;
{
#ifdef GDBM
	char name[BUFSIZ];
	(void) sprintf (name, "%s.gdbm", filename);
	filename = name;
	if ((thedb = gdbm_open (filename, 0, GDBM_WRITER, 0666, NULL)) == NULL) {
#else
# ifdef NDBM
	if((thedb = dbm_open(filename, O_RDWR, 0666)) == NULL) {
# else
	if(dbminit(filename) < 0) {
# endif
#endif
	    fprintf (stderr, "could not initialize data base '%s'", filename);
	    perror("");
	    exit(99);
	}
}

quit()
{
	exit(0);
}
